Commit code for Organization.

wpoosanguansit opened the pull request:
Commit code for Organization.

reid-spencer commented on organization/src/test/scala/com/improving/app/organization/utils/LoanedActorSystem.scala:
```diff
@@ -0,0 +1,28 @@
+trait LoanedActorSystem {
+  def configOverrides: Map[String, Any] = Map.empty
+
+  private val n = new AtomicInteger
```
Is this even needed ? It doesn't seem referenced in this trait.

wpoosanguansit replied:
It is used in the tests using trait CassandraTestContainer.

reid-spencer reviewed (commented):
Please review my feedback.

wpoosanguansit commented on build.sbt:
```diff
@@ -24,5 +26,6 @@ lazy val tenant = project.in(file("tenant"))
   .dependsOn(commonTypes % "compile->compile;test->test")
 
-(Compile / runMain) := (tenant / Compile / runMain).evaluated
-onLoad in Global := (onLoad in Global).value andThen (Command.process("project tenant", _))
```
Noted

wpoosanguansit commented on organization/src/main/resources/certs/ca.pem:
```diff
@@ -0,0 +1,15 @@
+-----BEGIN CERTIFICATE-----
```
I will remove this.

wpoosanguansit commented on organization/src/main/resources/certs/server1.key:
```diff
@@ -0,0 +1,16 @@
+-----BEGIN PRIVATE KEY-----
```
I will remove this.

wpoosanguansit commented on organization/src/main/scala/com/improving/app/organization/OrganizationServiceImpl.scala:
```diff
@@ -0,0 +1,339 @@
+      in,
+      {
+        case StatusReply.Success(
+              response @ Organization(_, _, _, _, _, _, _, _, _)
```
Noted.

wpoosanguansit commented on organization/src/main/scala/com/improving/app/organization/domain/Organization.scala:
```diff
@@ -1,43 +1,815 @@
+  }
+
+  private val commandHandler: (
+      OrganizationState,
```
I follow the pattern Shailendra did in member and I saw that it is being a state proto message which I am using it in the handler.  We do need to have the state proto message, right?

wpoosanguansit commented on organization/src/main/scala/com/improving/app/organization/utils/serialize/CborSerializable.scala:
```diff
@@ -0,0 +1,6 @@
+/** Marker trait for serialization with Jackson CBOR.
+  * Enabled in serialization.conf `akka.actor.serialization-bindings` (via application.conf).
+  */
+trait CborSerializable
```
Deleted.

wpoosanguansit commented on organization/src/main/scala/com/improving/app/organization/utils/serialize/AvroJacksonObjectMapperFactory.scala:
```diff
@@ -0,0 +1,40 @@
+
+class AvroJacksonObjectMapperFactory extends JacksonObjectMapperFactory {
+
+  override def newObjectMapper(bindingName: String, jsonFactory: JsonFactory): ObjectMapper = {
```
Deleted.

wpoosanguansit commented on organization/src/main/scala/com/improving/app/organization/utils/serialize/CborSerializable.scala:
```diff
@@ -0,0 +1,6 @@
+/** Marker trait for serialization with Jackson CBOR.
+  * Enabled in serialization.conf `akka.actor.serialization-bindings` (via application.conf).
+  */
+trait CborSerializable
```
Fixed.

wpoosanguansit commented on organization/src/main/scala/com/improving/app/organization/domain/Organization.scala:
```diff
@@ -1,43 +1,815 @@
+          OrganizationEvent,
+          OrganizationState
+        ](
+          persistenceId = PersistenceId(entityTypeHint, orgId),
```
Fixed.

wpoosanguansit commented on organization/src/main/scala/com/improving/app/organization/OrganizationServer.scala:
```diff
@@ -0,0 +1,116 @@
+      OrganizationServiceHandler(new OrganizationServiceImpl())
+
+    val bound: Future[Http.ServerBinding] = Http(system)
+      .newServerAt(interface = "127.0.0.1", port = 8080)
```
Fixed.

wpoosanguansit commented on organization/src/main/scala/com/improving/app/organization/OrganizationServer.scala:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  private def readPrivateKeyPem(): String =
+    Source.fromResource("certs/server1.key").mkString
```
Deleted.

wpoosanguansit commented on organization/src/main/scala/com/improving/app/organization/OrganizationServer.scala:
```diff
@@ -0,0 +1,116 @@
+      )
+
+    val bound: Future[Http.ServerBinding] = Http(system)
+      .newServerAt(interface = "127.0.0.1", port = 8080)
```
Fixed.

wpoosanguansit commented on organization/src/main/resources/grpc.conf:
```diff
@@ -0,0 +1,11 @@
+  grpc {
+    # consider setting this to a specific interface for your environment
+    interface = "0.0.0.0"
+    port = 8101
```
Fixed.

wpoosanguansit commented on organization/src/main/resources/cluster.conf:
```diff
@@ -0,0 +1,32 @@
+  }
+}
+
+akka.management {
```
Fixed.

wpoosanguansit commented:
The update on Mar 22 contains the following:

- Added organization by member and organization by owner projections
- Added specs for the projections

Fixes for feedback:

- Undo sbt project run snippet
- Remove cert and http start up
- Changed ports and Interface are variable to be set
- Changed TerminationDeadLine to be from config
- Changed match case of response to typed response
- Update extract id to be extractOrganizationId
- Changed request match cases to pass on request for processing
- Removed old serialize util

Note:

The integration test for projection does not write to the db yet.  It is being ignored for commit.
The projections are partially implemented as it is not allowed in Cassandra to delete records base on partial keys.
More work is needed to manually query the composite key data to perform the delete cleanly, ie. get all memberId from the organizations with certain orgId.

AlexWeinstein92 commented on ddl-scripts/create-organization-tables.sql:
```diff
@@ -0,0 +1,164 @@
+
+WITH insert_organization AS (
+INSERT INTO organization (org_id,
+		info_id,
```
Here you are inserting `info_id` into `organization` table, but `info_id` was not declared as a field in `organization` table. The same applies to the `meta_id` field. Please ensure there are unit or integration tests to test that this is really working.

wpoosanguansit replied:
I think I should remove this altogether as this was for the data model for relational db which is not used.  Cassandra is used for the projections.

AlexWeinstein92 commented on ddl-scripts/create-organization-tables.sql:
```diff
@@ -0,0 +1,164 @@
+--    address_id uuid,
+--    FOREIGN KEY (address_id) REFERENCES address(address_id),
+--    FOREIGN KEY (tenant_id) REFERENCES tenant(tenant_id),
+    FOREIGN KEY (org_id) REFERENCES organization(org_id),
```
This seems like it duplicates the `org_id` field on line 14, but I may just not be familiar with the syntax here

AlexWeinstein92 commented on ddl-scripts/create-organization-tables.sql:
```diff
@@ -0,0 +1,164 @@
+    url VARCHAR(255),
+    logo VARCHAR(255),
+    tenant_id uuid,
+--    address_id uuid,
```
Why is this commented out when `address_id` is still being inserted here below?

AlexWeinstein92 commented on ddl-scripts/create-organization-tables.sql:
```diff
@@ -0,0 +1,164 @@
+    UNIQUE(org_id)
+);
+
+WITH insert_organization AS (
```
I don't think we want to do this insertion in deployment.

AlexWeinstein92 commented on organization/ddl-scripts/create_tables.cql:
```diff
@@ -0,0 +1,96 @@
+-- This CQL script will create the keyspace and all tables needed for the this sample. 
```
Please include a link to wherever this code was taken from in a comment or other documentation. Its important to keep references like that easily accessible. Please do the same for any other files adapted for our use from other projects

AlexWeinstein92 commented on organization/src/main/scala/com/improving/app/organization/Main.scala:
```diff
@@ -1,14 +1,65 @@
+    val session = CassandraSessionRegistry(system).sessionFor(
+      "akka.persistence.cassandra"
+    )
+    // use same keyspace for the item_popularity table as the offset store
```
This comment seems like it comes from the framework you adapted from to write this file. Please either take it out or make it up to date with your adaptations.

AlexWeinstein92 commented on organization/src/main/scala/com/improving/app/organization/repository/OrganizationByMemberProjectionHandler.scala:
```diff
@@ -0,0 +1,303 @@
+                organization.orgId.map(_.id).getOrElse("OrgId is NOT FOUND."),
+                member.id,
+                organization.copy(
+                  members = organization.members ++ newMembers,
```
Most of the code here can be refactored into a single function - feel free to change the name of `handleEvent` to a better one if desired:

```
def handleEvent(newOrg: Organization) = repo
          .getOrganizationsByMemberByOrgId(orgId.id)
          .map(organizations => {
            Future.sequence(for {
              organization <- organizations
              member <- newMembers
            } yield {
              repo.updateOrganizationByMember(
                organization.orgId.map(_.id).getOrElse("OrgId is NOT FOUND."),
                member.id,
                newOrg
              )
            })
          })
          .map(_ => Done)

case MembersAddedToOrganization(Some(orgId), newMembers, actingMember, _) => {
  handleEvent(organization.copy(
    members = organization.members ++ newMembers,
    meta = Some(updateMetaInfo(organization.getMeta, actingMember))
  ))
}
```

dag-yoppworks commented:
This PR will not be merged - the branch became too much to review and transfer to new developers (👋 ), and it was not close to ready for merge. There are probably useful bits to draw on in this branch, if they can be found.

This is a cautionary tale about regularly committing, reviewing, and merging your code.
![alwaysbecommitting](https://user-images.githubusercontent.com/127801986/232582543-a2b81a40-a647-472c-bf2e-65a70e7811e4.png)
